[AArch64][GlobalISel] Fix lowering of i64->f32 itofp.

davemgreen opened the pull request:
This is a GISel equivalent of #130665, preventing a double-rounding issue in sitofp/uitofp by scalarizing i64->f32 converts. Most of the changes are made in the ActionDefinitionsBuilder for G_SITOFP/G_UITOFP. Because it is legal to convert i64->f16 itofp without double-rounding, but not a fpround f64->f16, that variant is lowered to build the two extends.

arsenm commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -7609,6 +7609,22 @@ LegalizerHelper::lowerU64ToF64BitFloatOps(MachineInstr &MI) {
+/// convert fpround f64->f16 without double-rounding, so we manually perform the
+/// lowering here where we know it is valid.
+static LegalizerHelper::LegalizeResult
+loweri64tof16ITOFP(MachineInstr &MI, Register Dst, LLT DstTy, Register Src,
```
Why is this under lower and not narrow scalar?

davemgreen replied:
Yeah I was wondering which was better. It felt more like a lower than a generic narrow. If a target wanted a single fptrunc (because they had an instruction for it), then they wouldn't want it to be split. So they would opt for the narrow. As far as I understand the narrows don't make extra legalization queries, there isn't a way to communicate two different types to them, and the alternative of a target hook doesn't sound the best. Maybe I'm missing something about how this should work.

arsenm replied:
The promote has a type parameter that tells the new type to use. I don't see what information is missing?

davemgreen replied:
AFAIR: We are trying to lower a `vNi64->vNf16 uitofp`, without scalarizing. For example a `v4i64->v4f16 uitofp` turns into
```
ucvtf v0.2d, v0.2d
ucvtf v1.2d, v1.2d
fcvtn v0.2s, v0.2d
fcvtn2 v0.4s, v1.2d
fcvtn v0.4h, v0.4s
```
So a `v4f16 fptrunc(v4f32 fptrunc(v4f64 uitofp(v4i64)))` (where the v4i64 is two vectors). If we initially say "widen type0 to f64" we get `v4f16 fptrunc(v4f64 uitofp(v4i64))` and the `v4f16 fptrunc(v4f64` needs to scalarize. If we say "widen type0 to f32" we get `v4f16 fptrunc(v4f32 uitofp(v4i64))` and the `v4f32 uitofp(v4i64` needs to scalarize (the point of this patch). That is what I meant by "there isn't a way to communicate two different types", you can't give it "widen to f64 via f32 please".

Note: for scalar we want it to become `f16 fptrunc(f64 uitofp(i64))` because we have an instruction for f64->f16 fptrunc. We don't want it to become two fptruncs.

(Whilst thinking about this, I wondered if we could lower to `uqxtn+uqxtn2+scvtf+fcvtn` for fp16 instead, but that would involve a TRUNCATE_USAT_U that we don't have for globalisel yet).

Currently we have it that lower means legalize via two steps, widen means lower directly, which might not be the best but fits into "lower does something different". What was the concrete suggestion for how this should work?

davemgreen replied:
If the answer is "Check that the G_FPTRUNC is not legal in widenScalar for G_UITOFP" then I can give that a go, we just don't call other legalization actions in most of LegalizerHelper at the moment. The vector types will need to split even if the vector types are not legal (wider vectors etc).

pranavk reviewed (approved):
LGTM

pranavk commented:
Test file will conflict after #130665 lands.

arsenm reviewed (changes requested):
Not sure why this is in lowering and not the promote/narrow cases

aemerson commented:
Why do we need to scalarize the vector case for GlobalISel but not for SDAG?

davemgreen committed:
[AArch64][GlobalISel] Fix lowering of i64->f32 itofp.

This is a GISel equivalent of #130665, preventing a double-rounding issue in
sitofp/uitofp by scalarizing i64->f32 converts. Most of the changes are made in
the ActionDefinitionsBuilder for G_SITOFP/G_UITOFP. Because it is legal to
convert i64->f16 itofp without double-rounding, but not a fpround i64->f16,
that variant is lowered to build the two extends.

davemgreen commented:
> Why do we need to scalarize the vector case for GlobalISel but not for SDAG?

As the commit message says this is a GISel equivalent of #130665. I will give it a rebase to update the tests.

davemgreen commented:
This came up again. Lets get it in to get it fixed.
